lib/puppet/data_providers/data_adapter.rb

```diff
@@ -28,15 +28,20 @@ def env_provider
     @env_provider ||= initialize_env_provider
   end
 
-  def module_provider(module_name)
-    @data[module_name] ||= initialize_module_provider(module_name)
+  def module_provider(module_name, nil_if_no_service=false)
+    # Test if the key is present for the given module_name. It might be there even if the
+    # value is nil (which indicates that no module provider is configured for the given name)
+    unless @data.include?(module_name)
+      @data[module_name] = initialize_module_provider(module_name, nil_if_no_service)
+    end
+    @data[module_name]
   end
 
   def self.create_adapter(environment)
     new(environment)
   end
 
-  def initialize_module_provider(module_name)
+  def initialize_module_provider(module_name, nil_if_no_service)
```

Comment:
nil_if_no_service is not needed - there is a 'none' provider for both modules and environments